### ISSUE TITLE
How to get composer sources?

### ISSUE THREAD

**Issue by OnkelTem** [composer/composer#6603]
to debug it.

Why there is no information on your MAIN page?

**Comment by stof** [comment]
the main page https://getcomposer.org/ has a big `Github` button linking to the github repo, which is where the source are

**Comment by OnkelTem** [comment]
@stof  Thanks for a quick reply, really appreciated this! I mean - how to get composer sources into the `vendor` dir? What dependency should I add to get it there? I need to debug Composer itself since something goes wrong with it when it fetches from git repos.

**Comment by curry684** [comment]
You cannot invoke Composer as a vendor dependency. Just `git clone https://github.com/composer/composer.git` and run `bin/composer <options>` in your IDE to debug (or fork first if you are willing to contribute a fix for the issue you're having).

**Comment by OnkelTem** [comment]
I don't know guys what did you mean, but I've figured out how to do this.

You just need to add these lines to you `composer.json`:

```json
    "require-dev": {
        "composer/composer": "*"
    }
```

and invoke composer from `./vendor/bin/composer` to get working debuggable development environment. You're kind of tough on providing clear answer guys.

**Comment by curry684** [comment]
No matter what the intent behind your original question was - your 'solution' is the most convoluted and fundamentally wrong method to go about it. It will work at first, and then you'll get stuck on the next steps again when you try to submit a fix for whatever issue you find.

To debug Composer you should clone the project sources and run `bin/composer <options>` in your IDE. If you don't understand how to do that, read up on the basics of Git and web development until you do. Don't blame us for giving the right answer, and definitely please do not frustrate the support process by advising other people to ignore good answers as well.

**Comment by OnkelTem** [comment]
> No matter what the intent behind your original question was - 

To debug it, I wrote it in plain English.

> your 'solution' is the most convoluted and fundamentally wrong method to go about it. 

I don't think so, it works pretty very well for me.

> It will work at first, and then you'll get stuck on the next steps again when you try to submit a fix for whatever issue you find.

I didn't ask about submitting fixes, I asked about getting sources to be able to see how things work. 

> To debug Composer you should clone the project sources and run bin/composer <options> in your IDE. 

Probably, but not necessary, as I'm fine with `export COMPOSER_ALLOW_XDEBUG = true`

> If you don't understand how to do that, 

No I do

> read up on the basics of Git and web development until you do. 

Give advises to those who asks them. I asked a simple question and you didn't provide an answer to it.

> ... and definitely please do not frustrate the support process by advising other people to ignore good answers as well.

I don't think that your answer is "better" them mine. Going *your way* one gets some packaged thing  which is not developer-friendly approach as it behaves like a black box. For this reason your advice is good for "start using composer" and for production, while mine - for a developer. I'd summarize: both our answers are correct, but you prefer to act like a dick.

**Comment by alcohol** [comment]
> Give advises to those who asks them. I asked a simple question and you didn't provide an answer to it.

> You cannot invoke Composer as a vendor dependency. Just git clone https://github.com/composer/composer.git and run bin/composer <options> in your IDE to debug (or fork first if you are willing to contribute a fix for the issue you're having).

While the above is not true (in the sense that 'cannot' should rather be 'should not'), it did answer your original question.

**Comment by curry684** [comment]
> but you prefer to act like a dick

...said the guy ignoring free support on how to debug software from someone with a "Contributor" badge next to their name proving they actually know how to best debug said software.

If you are going to ignore good answers anyway, please stay on Stack Overflow where that's normal and don't bother us on the development issue tracker.

**Comment by Seldaek** [comment]
Thanks everyone for your contributions to this healthy conversation. If you want to fling crap at each other please do it somewhere else.
